### ISSUE TITLE
.env files are ignored

### ISSUE THREAD

**Issue by ranisalt** [parcel-bundler/parcel#4110]
# 🐛 bug report

I'm trying to provide environment variables to select a couple of values on build time, however the env files ([as described here](https://parceljs.org/env.html)) are being ignored.

## 🎛 Configuration (.babelrc, package.json, cli command)

.env
```
PROJECT_PACKAGE=myproject
```

.env.development
```
API_ENDPOINT=localhost
```

config.ts
```ts
export default {
	project_name:
		process.env.PROJECT_PACKAGE === 'myproject'
			? 'My Project'
			: 'Not My Project',

	api_url:
		process.env.API_ENDPOINT === 'localhost'
			? 'http://localhost:8074'
			: 'https://api.example.com',
}
```

## 🤔 Expected Behavior
`config.ts` should export 
```ts
{
  project_name: 'My Project',
  api_url: 'http://localhost:8074'
}
```

## 😯 Current Behavior
`config.ts` exports
```ts
{
  project_name: 'Not My Project',
  api_url: 'https://api.example.com'
}
```

Upon further inspection (inserted `debugger` inside config.ts) I see that `process.env.NODE_ENV` is set to `development` (as expected), but `process.env.API_ENDPOINT` and `process.env.PROJECT_PACKAGE` are both `undefined`.

## 💻 Code Sample

See above

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 2.0.0-alpha.3
| Node             | 13.8
| npm/Yarn         | yarn 1.22
| Operating System | Arch Linux

**Comment by cdll** [comment]
i used to config my env configure in npm hooks like: 
````json
// package.json
{
  ...
  "scripts": {
    "start": "APP_ENV=local parcel serve",
    "build:dev": "APP_ENV=develop parcel build",
    "build:prod": "APP_ENV=production parcel build"
  },
  ...
}
````
wish some helps for u~

**Comment by ranisalt** [comment]
@cdll I have several more env vars so that would be very annoying to do.

**Comment by ranisalt** [comment]
The problem is parcel reads the env vars from the root of the package and not the packages inside this monorepo. I'll try to make a PR and fix this.

**Comment by mischnic** [comment]
Related: https://github.com/parcel-bundler/parcel/issues/3942#issuecomment-569084383, https://github.com/parcel-bundler/parcel/issues/3942#issuecomment-571192655

**Comment by ranisalt** [comment]
Yeah that's probably the same issue to those comments. It would be good for parcel to read all env files from the monorepo root down to the package dir, updating values as it travels down.

Or should the root .env.local be more important than the package .env.local?

**Comment by mischnic** [comment]
> Or should the root .env.local be more important than the package .env.local?

I'd say no.

**Comment by mjfneto** [comment]
Hi! I'm trying to deploy a project bundled with Parcel using Netlify. I am following Robert Cooper's [instructions](https://www.robertcooper.me/front-end-javascript-environment-variables):

>Specifying Environments
When running the normal parcel serve command (e.g. parcel index.html), the environment (specified by the `NODE_ENV` environment variable) will be development by default. To specify an environment other than development while using the parcel serve command, you can specify the environment when running the parcel serve command:

> `NODE_ENV=staging parcel index.html`

> When running the parcel build command (e.g. parcel build index.html), the environment will be production by default. Again, you can specify a specific environment when running the build command:

> `NODE_ENV=staging parcel build index.html`

And:

> You should commit all your `.env` files to source control with the exception of `.env*.local` files. The local `.env` files should only be used to tweak your app's configurations when running the app locally.

I am currently not commiting the `.env` file but I'm using the environment specification mentioned above. It's not providing access to staging variables during build. Is this due to the bug in discussion?

**Comment by mtxr** [comment]
what's the current status of this issue? Seems very important to still open.

**Comment by alessandrojcm** [comment]
Same here. In my case I'm trying to use `NODE_ENV` to do a simple conditional import:
```typescript
async function renderApp() {
    if (process.env.NODE_ENV === 'development') {
        // eslint-disable-next-line no-console
        await import('stuff');
    }
    render(<App />, document.getElementById('root'));
}
```
I'm using `cross-env` to set `NODE_ENV`, as so: `cross-env NODE_ENV=development parcel ./src/index.html`. But it doesn't work; moreover, if I console log `process.env.NODE_ENV` it prints out `production` :question:. Also tried putting a `.env.development` but it doesn't get picked up by parcel.

Using latest version (2.0.0-beta1).

**Comment by mischnic** [comment]
@alessandrojcm You don't need to specify NODE_ENV explicitly in your case. `parcel build` automatically sets `production` and in watch mode, `development`. Please provide a full example.

**Comment by alessandrojcm** [comment]
@mischnic Yes, I read that in the docs; I added the `cross-env` because for some reason it is not working in my project. Now, in [this](https://codesandbox.io/s/parcel-typescript-babel-plugin-styled-components-r4ho4) codesandbox it works as intended; but for some reason, in my project, it doesn't. You can check my project [here](https://github.com/alessandrojcm/images-of-the-world/tree/develop).

**Comment by desfero** [comment]
Stumbled upon the same problem today. Would be good to provide some options to overwrite `projectRoot` to be related to the given package `parcel` is used for in monorepo setup. 
`projectRoot` that points to the root is causing also problems at least with babel (see #7075)

**Comment by rafaelnai** [comment]
In my case, i solved it by adding the entrie file path in package.json.

```json
"scripts": {
    "start": "parcel",
    "build": "parcel build src/index.html"
 }
```

**Comment by charIeszhao** [comment]
> The problem is parcel reads the env vars from the root of the package and not the packages inside this monorepo. I'll try to make a PR and fix this.

Hey, is this still going to happen? Just came across this issue last night and I was really scratching my head figuring this out. 
Reading .env files from sub packages sounds more reasonable to me. Thanks! @ranisalt

**Comment by hnordt** [comment]
> In my case, i solved it by adding the entrie file path in package.json.
> 
> ```json
> "scripts": {
>     "start": "parcel",
>     "build": "parcel build src/index.html"
>  }
> ```

I tried this, but it didn't work in my case.

**Comment by beefs** [comment]
> In my case, i solved it by adding the entrie file path in package.json.
> 
> ```json
> "scripts": {
>     "start": "parcel",
>     "build": "parcel build src/index.html"
>  }
> ```

This works for me too thanks!!!

**Comment by rubixibuc** [comment]
I just ran into this problem as well. I will have to completely stop using parcel for any projects if this isn't updated.

**Comment by mischnic** [comment]
> In my case, i solved it by adding the entrie file path in package.json.

This should be fixed now in the latest nightly version by https://github.com/parcel-bundler/parcel/pull/7537

Try using `yarn add parcel@nightly`

**Comment by d-lach-aliz** [comment]
@mischnic, @devongovett  can you confirm that 
> The problem is parcel reads the env vars from the root of the package and not the packages inside this monorepo. I'll try to make a PR and fix this.

is fixed?

It still does not work for me, testing with build ^2.0.0-nightly.1071
by not working I mean that only .env files in the root of the project are loaded, .envs in monorepo packages are ignored.

**Comment by ghannamr** [comment]
I'm also still having trouble loading `.env.local` from my monorepo package. Loading the env variables only works when the `.env.local` file is placed in the root of the project.

I'm using the latest 2.6.0

**Comment by tomtobac** [comment]
I am using `parcel@2.6.0` and I am having the same issue. Monorepo with multiple `.env` in each package

**Comment by mischnic** [comment]
Yes, currently it only searches for dotfiles in the project (=monorepo) root

https://github.com/parcel-bundler/parcel/blob/3b3b13557dba9d1a403c86550915c6013dd78dfe/packages/core/core/src/loadDotEnv.js#L18-L35

**Comment by havenbarnes** [comment]
Spent a couple days trying to solve this for myself. I tried using `dot-env` and `env-cmd` directly to reference the correct env files instead of letting parcel look for them in the project root - no dice for me at least.

The only feasible "workaround" I stumbled upon was the one mentioned in here https://github.com/parcel-bundler/parcel/issues/4264, which is to use `cross-env`. Of course, you're not using `.env` files at that point, but it can get the job done depending on how many variables you need to set.

**Comment by havenbarnes** [comment]
A pretty easy way to use the `cross-env` workaround and still use your .env files is to do 
```
cross-env $(cat .env.development) parcel ...
```

**Comment by khusamov** [comment]
Are they planning to fix this bug?

**Comment by jraoult** [comment]
I just want to point out that the root cause of this issue is being discussed here: https://github.com/parcel-bundler/parcel/issues/7579 and that[ this workaround](https://github.com/parcel-bundler/parcel/issues/7579#issuecomment-1080355106) while imperfect is doing the job (mostly).

**Comment by svaterlaus** [comment]
For anyone unsatisfied with the solutions offered so far, an easy fix for me is using [dotenv-cli](https://www.npmjs.com/package/dotenv-cli) and running Parcel like this:

```
dotenv -- parcel src/index.html --port 3001
```

It loads the environment variables from the nearest `.env` file instead of only from the monorepo root.
